### ISSUE TITLE
Having multiple (many) Jetpack Compose tests causes android.support.test.espresso.AppNotIdleException for some tests

### ISSUE THREAD

**Issue by alwa** [robolectric/robolectric#7055]
### Description
When using different effects, such as `DisposableEffect` or `LaunchedEffect`, declared in a composable to be tested I have noticed erratic behavior by Robolectric. I also see issues for some tests even without use of effects, but removing the effects made some of my tests pass at least.

The tests may run fine locally (Windows) but fail in Gitlab (android.support.test.espresso.AppNotIdleException) on the UI rendering step.

### Steps to Reproduce
My code could look like this:
```
MyComposablesKt {
@Composable
fun MyComposable() {
   // Declaration of effect + normal composable UI code
}

```

My tests look like this:
```
@Config(sdk = [Build.VERSION_CODES.O_MR1])
@RunWith(AndroidJUnit4::class)
@LooperMode(LooperMode.Mode.PAUSED)
class MyComposablesKtTest {

    @get:Rule
    val composeTestRule = createComposeRule()

    private val myActionButton get() = composeTestRule.onNodeWithContentDescription("Action Button")

    @Test
    fun `MyComposable - when button clicked - does something`() {
        val navController: NavController = mockk(relaxed = true)
        composeTestRule.setContent {
            MyComposable()
        }
        myActionButton.performClick()
        ShadowLooper.runUiThreadTasks()

       // Assertions
    }

}
```

The test fails on this line:
```
composeTestRule.setContent {
```

### Robolectric & Android Version
Robolectric 4.7.3
AGP 7.1.1
Gradle 7.4
androidx-test-junit 1.1.4-alpha03
androidx-test-monitor 1.5.0
androidx-test-runner 1.4.1-alpha03

**Comment by alwa** [comment]
It's possible, or even likely, that there is another root cause for these problems. 
For example this test fails as well (on line with last assertion):

```
    fun `CloseDialog - when showDialog true and close button clicked - dismisses dialog`() {
        composeTestRule.setContent {
            CloseDialog(
                showDialog = mutableStateOf(true),
                contentDescription = "expected value",
                titleText = "",
                bodyText = ""
            )
        }
        ShadowLooper.runUiThreadTasks()

        closeButtonNode.performClick()

        composeTestRule.onNodeWithContentDescription("expected value").assertDoesNotExist()
    }
```

Prod code:

```
@Composable
fun CloseDialog(
    onDismiss: () -> Unit = { showDialog.value = false },
    showDialog: MutableState<Boolean>,
    contentDescription: String,
    titleText: String,
    bodyText: String
) {
    if (showDialog.value) {
        AlertDialog(modifier = Modifier.semantics { this.contentDescription = contentDescription }, title = {
            Text(...)
        }, text = {
            Text(...)
        }, onDismissRequest = { onDismiss.invoke() },
            confirmButton = {
                TextButton(onClick = { onDismiss.invoke() }) {
                    Text(text = "Close")
                }
            })
    }
}

```

**Comment by alwa** [comment]
It seems like the culprit is:

```
@Config(sdk = [Build.VERSION_CODES.O_MR1])
@RunWith(AndroidJUnit4::class)
@LooperMode(LooperMode.Mode.PAUSED)
```

Replacing it with only

```
@RunWith(AndroidJUnit4::class)
```

seems to work better

**Comment by hoisie** [comment]
Have you also tried:

[ShadowLooper .runUiThreadTasksIncludingDelayedTasks](https://source.corp.google.com/piper///depot/google3/third_party/java_src/robolectric/shadows/framework/src/main/java/org/robolectric/shadows/ShadowLooper.java;bpv=1;bpt=1;l=199?q=shadowlooper&sq=package:piper%20file:%2F%2Fdepot%2Fgoogle3%20-file:google3%2Fexperimental&gsn=runUiThreadTasksIncludingDelayedTasks&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgoogle3%3Flang%3Djava%3Fpath%3Dorg.robolectric.shadows.ShadowLooper%2340baa3858486de9fd36ea4c29e97731896061911446a2bd0cd0a60d63ba7041c)() {

**Comment by alwa** [comment]
> When `@LooperMode(LooperMode.Mode.PAUSED)` is removed, is the test running in LEGACY looper mode? What is the value of ShadowLooper.looperMode()?

Hmm, according to javadoc it should default to LooperMode.Mode.PAUSED, which is weird as it is the same value that was not working well for me. On the other hand I still see this exception sometimes even without @LooperMode set.
So this is still a mystery.

**Comment by alwa** [comment]
> Have you also tried ShadowLooper.runUiThreadTasksIncludingDelayedTasks?

I'm pretty sure I have tried that as well

**Comment by hoisie** [comment]
Sounds like a race condition. Maybe compose effects are occurring on a background thread, and it's possible that the result are seeing non-deterministic given that some processing is being done in a separate thread.

@jfresen do you have any thoughts on why this may be happening?

**Comment by alwa** [comment]
> Sounds like a race condition. Maybe compose effects are occurring on a background thread

I have seen this behavior even when testing composables without effects, but having effects in the composable will definitely cause the issue

**Comment by hoisie** [comment]
Do you have a sample app that demonstrates the problem? Is there a stack trace that you're seeing? I am unfortunately not too experienced with Compose, but I can try to ping someone from the Compose team to take a closer.

**Comment by jfresen** [comment]
Compose effects are fully controlled by the MainTestClock, so there should be no possibility for a race there.

> The test fails on this line:
> 
> ```
> composeTestRule.setContent {
> ```

What is the failure message / stacktrace that you see here? setContent performs an initial composition immediately (subject to attached to window), so I'm keen to know where it fails

**Comment by alwa** [comment]
> What is the failure message / stacktrace that you see here? setContent performs an initial composition immediately (subject to attached to window), so I'm keen to know where it fails

```
    androidx.test.espresso.AppNotIdleException: Compose did not get idle after 1185356 attempts in 60 SECONDS. Please check your measure/layout lambdas, they may be causing an infinite composition loop. Or set Espresso's master idling policy if you require a longer timeout. The following Idle Conditions failed .
        at androidx.test.espresso.AppNotIdleException.create(AppNotIdleException.java:11)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:55)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:48)
        at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.runOnUiThread(AndroidSynchronization.android.kt:33)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy.runUntilIdle(RobolectricIdlingStrategy.android.kt:48)
        at androidx.compose.ui.test.AndroidComposeTest.waitForIdle(ComposeTest.android.kt:169)
        at androidx.compose.ui.test.AndroidComposeTest.waitForIdle(ComposeTest.android.kt:162)
        at androidx.compose.ui.test.AndroidComposeTest.setContent(ComposeTest.android.kt:157)
        at androidx.compose.ui.test.junit4.AndroidComposeTestRule.setContent(AndroidComposeTestRule.android.kt:197)
```

**Comment by jfresen** [comment]
I see. That could be an indication that you have a self-invalidating loop somewhere in your code. Could you try and add an apply observer to your test like in the below gist to find out if there are state variables that are continuously written?

https://gist.github.com/jfresen/44f3838fbd00e3c4342c7be4c9a4e6f7

**Comment by alwa** [comment]
I slightly modified the snippet above to print everything to standard err. Here is the output from one of the failing tests.
To me, after a quick look, it's not clear if this indicates that something is wrong with the composable... 

```
MyComposableKtTest > Test name
STANDARD_ERROR
    [0] applying snapshot
    [1] object MutableState(value=androidx.compose.ui.platform.AndroidComposeView$ViewTreeOwners@3f26a6d7)@74557308 modified
    [2] applying snapshot
    [3] object androidx.compose.runtime.snapshots.SnapshotStateList@6214d835 modified
    [4] object MutableState(value=Inactive)@362697403 modified
    [5] object MutableState(value=false)@1032649273 modified
    [6] object MutableState(value=Inactive)@2103563641 modified
    [7] object androidx.compose.runtime.snapshots.SnapshotStateList@622b5813 modified
    [8] object MutableState(value=false)@1264957676 modified
    [9] object MutableState(value=false)@248682040 modified
    [10] object androidx.compose.runtime.snapshots.SnapshotStateList@6b356e69 modified
    [11] object MutableState(value=Inactive)@913979607 modified
    [12] applying snapshot
    [13] object MutableState(value=androidx.navigation.NavBackStackEntry@2628f8db)@938474505 modified
    [14] object MutableState(value=true)@1654280726 modified
    [15] applying snapshot
    [16] object MutableState(value=[])@1599462721 modified
    [17] object MutableState(value=false)@2129474551 modified
    [18] object androidx.compose.runtime.snapshots.SnapshotStateList@4278c3c5 modified
    [19] applying snapshot
    [20] object MutableState(value=false)@1461036661 modified
    [21] object MutableState(value=1.0)@1999562476 modified
    [22] object MutableState(value=false)@1286591143 modified
    [23] object MutableState(value=false)@633014355 modified
    [24] object MutableState(value=Size(320.0, 414.0))@158352541 modified
    [25] object MutableState(value=-9223372036854775808)@1614093338 modified
    [26] object MutableState(value=false)@1195037141 modified
    [27] object MutableState(value=-9223372036854775808)@1408936330 modified
    [28] object MutableState(value=false)@382978369 modified
    [29] object MutableState(value=-9223372036854775808)@2097231314 modified
    [30] object MutableState(value=false)@903957428 modified
    [31] object MutableState(value=-9223372036854775808)@430353695 modified
    [32] object MutableState(value=false)@395021558 modified
    [33] object MutableState(value=false)@1208970769 modified
    [34] applying snapshot
    [35] object MutableState(value=470.0)@1254717757 modified
    [36] object MutableState(value=470.0)@[4268](https://myserver/-/jobs/323960#L4268)19723 modified
    [37] applying snapshot
    [38] object MutableState(value=Function2<java.lang.Float, java.lang.Float, java.lang.Float>)@1150922136 modified
    [39] object MutableState(value=125.0)@1497990351 modified
    [40] object MutableState(value={470.0=Hidden, 469.0=Expanded})@772421810 modified
    [41] applying snapshot
    [42] object MutableState(value=() -> kotlin.Boolean)@2104847459 modified
    [43] object MutableState(value=androidx.compose.foundation.gestures.DragLogic@705e345e)@1332301023 modified
    [44] applying snapshot
    [45] object MutableState(value=() -> kotlin.Boolean)@2104847459 modified
    [46] object MutableState(value=androidx.compose.foundation.gestures.DragLogic@220fed50)@1332301023 modified
MyComposableKtTest > Test name FAILED
    androidx.test.espresso.AppNotIdleException: Compose did not get idle after 1177132 attempts in 60 SECONDS. Please check your measure/layout lambdas, they may be causing an infinite composition loop. Or set Espresso's master idling policy if you require a longer timeout. The following Idle Conditions failed .
        at androidx.test.espresso.AppNotIdleException.create(AppNotIdleException.java:11)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:55)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:48)
        at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.runOnUiThread(AndroidSynchronization.android.kt:33)
        at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy.runUntilIdle(RobolectricIdlingStrategy.android.kt:48)
        at androidx.compose.ui.test.AndroidComposeTest.waitForIdle(ComposeTest.android.kt:169)
        at androidx.compose.ui.test.AndroidComposeTest.waitForIdle(ComposeTest.android.kt:162)
        at androidx.compose.ui.test.AndroidComposeTest.setContent(ComposeTest.android.kt:157)
        at androidx.compose.ui.test.junit4.AndroidComposeTestRule.setContent(AndroidComposeTestRule.android.kt:197)
        at MyComposableKtTest.Test name(AppNavigationKtTest.kt:89)
```

**Comment by hoisie** [comment]
@alwa Would it be possible to create a small sample app on GitHub that repros this problem? I don't have much context into these compose stack traces so that would be super helpful to poke around this issue.

**Comment by alwa** [comment]
> @alwa Would it be possible to create a small sample app on GitHub that repros this problem? I don't have much context into these compose stack traces so that would be super helpful to poke around this issue.

I'll see what I can do. Seems like the issues only happen when multiple tests are run in parallell - which could make it tricky. But if I succeed in isolating the failure into smaller pieces of code I'll see if I can make a sample project.

**Comment by jfresen** [comment]
Hi, apologies for replying late. I don't see any obvious repeating pattern in the mutated state, so it looks like there's no self-invalidating loop - at least not state-wise.

Is it possible to add timestamps? I'm curious to know where the 60 seconds are spent.

You could try logging all activity on the main looper to see if it's busy looping, by using `Looper.setMessageLogging(Printer)`. Here timestamps could also be useful.

**Comment by alwa** [comment]
I'm sorry for the late reply.

> Is it possible to add timestamps? I'm curious to know where the 60 seconds are spent.

I'm not sure what code I was posting earlier but I found another affected test for a simple composable where I put different timestamps.

The code looks something like this (a bit shortened/simplified for readability):

```
@Composable
fun MyComposable(navController: NavController) {
    System.err.println("Top of composable ${System.currentTimeMillis()}")
    with(rememberSystemUiController()) {
        setSystemBarsColor(color = MaterialTheme.colors.background, isNavigationBarContrastEnforced = true)
        setNavigationBarColor(color = MaterialTheme.colors.inactive)
    }
    Scaffold(modifier = Modifier, topBar = {
        System.err.println("Top of topBar ${System.currentTimeMillis()}")
        TopAppBar(
            elevation = 0.dp, backgroundColor = Color.Transparent
        ) {
            // Some content
        }
    }) {
        System.err.println("Top of scaffold content ${System.currentTimeMillis()}")
        val scrollState = rememberScrollState()
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(scrollState)
        ) {
            System.err.println("Inside of column ${System.currentTimeMillis()}")
            // A bunch of Text composables
        }
    }
}
```

The logs show this:
```  
    [0] applying snapshot
    [2] object MutableState(value=androidx.compose.ui.platform.AndroidComposeView$ViewTreeOwners@1843bf7d)@992412377 modified
    Top of composable 1648066865984
    Top of topBar 1648066865992
    [4] applying snapshot
    [6] object MutableState(value=false)@1265686385 modified
    [8] object MutableState(value=Inactive)@625110462 modified
    Top of scaffold content 1648066866004
    Inside of column 1648066866004
    [10] applying snapshot
    [12] object MutableState(value=Inactive)@2078263231 modified
    [14] object MutableState(value=Inactive)@2029006241 modified
    [16] object MutableState(value=Inactive)@33622885 modified
    [18] object MutableState(value=Inactive)@2077238633 modified
    [20] object MutableState(value=Inactive)@971511986 modified
```  

and then:
![image](https://user-images.githubusercontent.com/1206247/159789663-6aa64230-5105-4b2d-9ebf-5806151cab36.png)

> 
> You could try logging all activity on the main looper to see if it's busy looping, by using `Looper.setMessageLogging(Printer)`. Here timestamps could also be useful.

`Looper.setMessageLogging(Printer)` doesn't compile. Do you mind elaborating on what needs to be done?

**Comment by jfresen** [comment]
> Looper.setMessageLogging(Printer) doesn't compile.
Well, you'll need to feed it a Printer

As for the timestamps, I was kinda hoping you'd replace your stderr output with logcat output.

Can you please provide a complete and minimal sample that reproduces the problem? Preferably a single test method that sets the compose content using the test rule's setContent method, so we have an isolated case.

**Comment by alwa** [comment]
> Well, you'll need to feed it a Printer
Pardon my ignorance. I never use loopers or printers, any gist for this?

> As for the timestamps, I was kinda hoping you'd replace your stderr output with logcat output, which formats timestamps in a readable way and which contains process and thread ids, which can sometimes be useful

@jfresen Unfortunately I don't see it in the test output if it's logged to logcat. I run the tests from command line.

> Can you please provide a complete and minimal sample that reproduces the problem? Preferably a single test method that sets the compose content using the test rule's setContent method, so we have an isolated case.

I can't reproduce the failure when running this test alone but this test always fails (100% reproducable) when running all tests (hundreds). So I suspect a memory leak in Robolectric/AndroidX test. So I'm afraid this approach would be a waste of time.

**Comment by jfresen** [comment]
> Pardon my ignorance. I never use loopers or printers, any gist for this?

To print to logcat: `Looper.getMainLooper().setMessageLogging { line -> Log.v("TEST_TAG (or anything that helps you identify the logcat messages)", line) }`
To print to stderr: `Looper.getMainLooper().setMessageLogging(PrintStreamPrinter(System.err))`

> @jfresen Unfortunately I don't see it in the test output if it's logged to logcat. I run the tests from command line.

logcat can be obtained from the command line with `adb logcat -d` after the test has run, or you can view it in Android Studio

> I can't reproduce the failure when running this test alone but this test always fails (100% reproducable) when running all tests (hundreds). So I suspect a memory leak in Robolectric/AndroidX test. So I'm afraid this approach would be a waste of time.

Oh, that changes things a lot. The issue seems to be with test isolation rather than an individual test then. You could try to find the minimum set of tests you need to run to trigger the failure by means of binary search.

Either way, if there's a bug in our frameworks, we really need a way to reproduce locally in order to help you and fix the bug on our side.

**Comment by alwa** [comment]
> To print to logcat: Looper.getMainLooper().setMessageLogging { line -> Log.v("TEST_TAG (or anything that helps you identify the logcat messages)", line) }
To print to stderr: Looper.getMainLooper().setMessageLogging(PrintStreamPrinter(System.err))

Thanks!

> logcat can be obtained from the command line with adb logcat -d after the test has run, or you can view it in Android Studio

I'm afraid I couldn't make it work with logcat (I can only make it work with a connected device). It looks like this:

![image](https://user-images.githubusercontent.com/1206247/159914827-d25b8535-ebda-40bb-9320-7554cb402adf.png)

> Oh, that changes things a lot. The issue seems to be with test isolation rather than an individual test then. You could try to find the minimum set of tests you need to run to trigger the failure by means of binary search. Either way, if there's a bug in our frameworks, we really need a way to reproduce locally in order to help you and fix the bug on our side.

Ok, got it. I'm not sure when I will find the time to do this but now I know at least 👍
Just for the sake of it I'm going to dump (relevant parts of) configurations in my project here. Maybe it's something obvious that I'm doing wrong in how the tests are set up that would lead to these problems:

build.gradle.kts:

```kotlin

   defaultConfig {
        minSdk = 28
        targetSdk = 31
        multiDexEnabled = true
        testInstrumentationRunner = "androidx.test.runner.AndroidJUnitRunner"
        vectorDrawables.useSupportLibrary = true
        buildConfigField("int", "MIN_SDK_VERSION", "$minSdk")
    }
    buildFeatures {
        compose = true
    }
    composeOptions {
        kotlinCompilerExtensionVersion = libs.versions.androidx.compose.get()
    }
    compileOptions {
        isCoreLibraryDesugaringEnabled = true
        sourceCompatibility = JavaVersion.VERSION_11
        targetCompatibility = JavaVersion.VERSION_11
    }
    kotlinOptions {
        jvmTarget = JavaVersion.VERSION_11.toString()
    }
    testOptions {
        unitTests.isIncludeAndroidResources = true
        unitTests.isReturnDefaultValues = true
        unitTests.all { test ->
            test.also { testTask ->
                testTask.maxHeapSize = "4g"
            }
            test.testLogging {
                events.addAll(listOf(PASSED, SKIPPED, FAILED))
                showCauses = true
                showExceptions = true
                exceptionFormat = FULL
            }
        }
    }
```

Test class header annotations:

```kotlin
@Config(sdk = [Build.VERSION_CODES.P])
@RunWith(AndroidJUnit4::class)
```

**Comment by Evin1-** [comment]
Had some trouble testing a suite with multiple Composables that were using a [ClickableText](https://developer.android.com/reference/kotlin/androidx/compose/foundation/text/package-summary#ClickableText(androidx.compose.ui.text.AnnotatedString,androidx.compose.ui.Modifier,androidx.compose.ui.text.TextStyle,kotlin.Boolean,androidx.compose.ui.text.style.TextOverflow,kotlin.Int,kotlin.Function1,kotlin.Function1)), which seems to be using a `LaunchedEffect` to handle gestures.

Adding `rule.mainClock.autoAdvance = false` before the setContent fixed it.

```kotlin
rule.mainClock.autoAdvance = false
rule.setContent { MyComposable() }
```

https://developer.android.com/jetpack/compose/testing#disable-autosync

**Comment by alwa** [comment]
> ```kotlin
> rule.mainClock.autoAdvance = false
> ```

Thanks for the suggestion. Looks like this setting could/should have impact. Unfortunately, in my case, I can still reproduce AppNotIdleException for tests that do this in the test setup method. As before, these same tests run successfully in isolation.

**Comment by alwa** [comment]
Note: A lot more of my compose tests break after I update these dependencies without altering any other code:

**Compose**: 1.2.0-alpha05 -> 1.2.0-alpha06
**Accompanist**: 0.24.4-alpha -> 0.24.5-alpha

Maybe this could be a clue to someone?

**Comment by alwa** [comment]
It looks like I have a breakthrough on this issue! More tests to perform before I can fully confirm but this is my theory for the root cause.

I have not declared an explicit dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-core** in the production code but I have a declared dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-test**. So what I suspect is that there was a version mismatch between the implicit dependency and the test dependency causing strange behavior.

The fix:
Make sure that any dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-core** uses the same version as **org.jetbrains.kotlinx:kotlinx-coroutines-test**

**Comment by jasco** [comment]
I observed the same kind of problem using compose 1.2.0-beta02.  I realize this is a little light on details without including my full dependency list.  I just wanted second the observation that there might be a race condition and also report that changing compose/compose-compiler/kotlin appeared to fix this particular failure, in my case.

My observations included
a. tests were working until merged with another branch with non-overlapping code
b. testing the class of problem tests in isolation passed
c. run as a suite, two tests in the problem class consistently timed out
d. removing blocks of tested but seemingly unrelated and benign code (no animations or anything apparently special) could allow the timing out tests to pass
e. removing an unrelated test that was testing the presence of an indeterminate progress would allow it to pass

The results were extremely repeatable for a given code base as long as I cleaned each time.  Oddly sometimes after running in isolation successfully, a suite run on its heels would sometimes succeed repeatedly.  I imagine this might have had to do with something being cached and running faster.  (I think in these cases I may have observed the test execution order had changed if the output in AndroidStudio is to be believed.)

Thinking the error might have been related to an animation in another test not shutting down quickly enough, I tried disabling animations but nothing had any effect and I am not sure how to check if it had the intended effect or how to get visibility to further examine or diagnose the behavior.

In my case kotlinx-coroutines-core and kotlinx-coroutines-test were already anchored at the same version.  Interestingly though the problem was consistently fixed when I changed compose to 1.1.1, compose-compiler 1.2.0 and kotlin 1.7.0.

While the change of dependencies fixed the test timeouts it also cause another test (previously working) to hang indefinitely.  The newly problematic test tests click actions in an action menu.  It seemed to hang on the performClick action.  After failing to find a solution to that problem, I resorted to ignoring the tests with plans to revisit in the future.

**Comment by vishnumj** [comment]
I'm also facing the exact same issue, the test runs fine locally, but fails in GitHub Actions. But, @alwa's solution is working for me but in a different way, the test function is passing when I annotate, `@Config(sdk = [Build.VERSION_CODES.P])` or `@Config(sdk = [Build.VERSION_CODES.Q])` 

🤷‍♂️

**Comment by alixwar** [comment]
> It looks like I have a breakthrough on this issue! More tests to perform before I can fully confirm but this is my theory for the root cause:
> 
> I have not declared an explicit dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-core** in the production code but I have a declared test dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-test**, as can be seen here: [#7055 (comment)](https://github.com/robolectric/robolectric/issues/7055#issuecomment-1077572803). So what I suspect is that there was a version mismatch between the implicit dependency and the test dependency causing flaky behavior.
> 
> The fix: Make sure that any implicit or explicit dependency to **org.jetbrains.kotlinx:kotlinx-coroutines-core** uses the same version as **org.jetbrains.kotlinx:kotlinx-coroutines-test**

This is me (@alwa) from my work account. This above is not solving the root cause, I'm afraid. I still see these errors.
Also I want to add that I still see the errors with the latest libraries (compose 1.3, AGP 7.3, Gradle 7.5)

**Comment by alixwar** [comment]
As I still have not found the root cause I want to at least share my thought so far.
I think the problem is coroutine-related. Maybe a memory leak building up delays somehow.

My general advice, at this current time, to avoid/circumvent this problem is to follow the guide on https://developer.android.com/kotlin/coroutines/coroutines-best-practices#inject-dispatchers and to avoid hard-coded dispatchers.

Example test class:
```
@Config(sdk = [BuildConfig.MIN_SDK_VERSION])
@RunWith(AndroidJUnit4::class)
@LooperMode(LooperMode.Mode.PAUSED)
class FooTest {

    private val testDispatcher = UnconfinedTestDispatcher()
    private val testScope = TestScope(testDispatcher)
    private val testContext: Context = ApplicationProvider.getApplicationContext()

    @get:Rule
    val composeTestRule = createComposeRule()
    
    @Before
    fun setUp() {
        Dispatchers.setMain(testDispatcher)
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `example test`() = testScope.runTest {
        viewModel = FooViewModel(
            context = ApplicationProvider.getApplicationContext(),
            ioDispatcher = testDispatcher
        )
        composeTestRule.setContent {
            MoreScreen(viewModel = viewModel)
        }

        // Assert stuff
    }
```

**Comment by moushao** [comment]
Hi， every bros, Unfortunately I also ran into this problem after I upgrade project version from 31 to 33!
compileSdk = 31 -> 33
targetSdk = 30 -> 33
minSdk = 24 -> 29
buildToolsVersion = "30.0.2" -> "33.0.3"

We never meet this issue until we updated。
like as @alwa said, I can't reproduce if I running test alone, but when I running all tests, it will 100% reproduce. I'm guessing if these tests are grabbing resources, or not releasing them.

I had tried all method what you talked, but not work. I will be creazy!

**Comment by alwa** [comment]
> Hi， every bros, Unfortunately I also ran into this problem after I upgrade project version from 31 to 33! compileSdk = 31 -> 33 targetSdk = 30 -> 33 minSdk = 24 -> 29 buildToolsVersion = "30.0.2" -> "33.0.3"
> 
> We never meet this issue until we updated。 like as @alwa said, I can't reproduce if I running test alone, but when I running all tests, it will 100% reproduce. I'm guessing if these tests are grabbing resources, or not releasing them.
> 
> I had tried all method what you talked, but not work. I will be creazy!

One more advice: Add test logging, like this (kotlin DSL example):

```kotlin
 testOptions {
        unitTests.all { test ->
            test.testLogging {
                events.addAll(listOf(PASSED, SKIPPED, FAILED))
                showCauses = true
                showExceptions = true
                exceptionFormat = FULL
            }
        }
    }
```

Now you can see in the build log what were the last tests to pass before `AppNotIdleException` happened. In many cases that is where you have a problem with the test.

Good luck!

**Comment by moushao** [comment]
hello @alwa , thanks for your suggestion very much!!! but our team also not fix this issue.

But I seem to have found some pattern，for example, all test which failed with the code:
`comoseTestRule.onNodeWitheText("xxxxxx").performClick()`
 after performClick，it seem liked the thread was suspended, or the idle resource not be released。
 
 but if I coding as below， it will be passed!
> `composeTestRule.mainClock.autoAdcance = false `
`composeTestRule.setContent{
           `ComposeView() `
}`
`testDispatcher.scheduler.apply{ advanceTimeBy(300); runCurrent()}`
`comoseTestRule.onNodeWitheText("xxxxxx").performClick()`

**Comment by SoJexy** [comment]
We are also running into a similar problem running multiple tests in parallel. When run in isolation they pass but ran in a suite we've had them fail. I can also confirm that adding `@Config(sdk = [Build.VERSION_CODES.Q])` fixes it for us.

**Comment by Eightyplus** [comment]
We have the same issue due to an old version androidx.fragment-testing (https://github.com/robolectric/robolectric/issues/6262). 
Our solution was to run on different config=sdks, which somehow isolations the tests from each other.
We cannot update to 1.4.1 yet due to Leanback. So closing the `ActivityScenario` solved the issue and we can run the tests on the same SDK. 

So yes, it seems this is either an actual warning or a listener or something that has not been shutdown correctly 🤯

**Comment by likeprogram** [comment]
I also encountered this issue, after investigation, found a way to reproduce this issue stably. I have created a [demo project ](https://github.com/likeprogram/RobolectricDemo) which can reproduce this issue stably.

In my scenario, this issue is caused by mixing compose test and view model test. For example, there are three test files

1. AScreenTest.kt, compose test, run with RobolectricTestRunner, will call setContent method, the default looper mode is PAUSED
2. AViewModelTest.kt, view model test, run with RobolectricTestRunner, will not call setContent but will modify mutable state, the default looper mode is PAUSED
3. BScreenTest.kt, compose test, run with RobolectricTestRunner, will call setContent method, the default looper mode is PAUSED

Please note the test must be executed as this order, let's explain the key phase when executing each test file

Execute AScreenTest.kt
1. When calling setContent for the first time, the GlobalSnapshotManager.ensureStarted will also be called which will setup a global snapshot write observer;
2. Afterwards, as long as there is a state modification, the observer will execute channel.trySend(Unit) to try to resume the suspended coroutine;
3. Then AndroidUiDispatcher.dispatch will be executed, as the scheduledTrampolineDispatch is false, a dispatchCallback will be posted to the main thread handler;
4. Once the main thread handle got executed, the posted dispatchCallback will be executed, then the suspended coroutine is resumed and the sendApplyNotifications is executed.
Everything is OK until now.

Execute AViewModelTest.kt
1. As the android SDK version and looper mode is the same as AScreenTest.kt, will reuse the previous Robolectric sandbox, will reuse the GlobalSnapshotManager and AndroidUiDispatcher;
2. As long as there is a state modification, the observer will execute channel.trySend(Unit) to try to resume the suspended coroutine;
3. Then AndroidUiDispatcher.dispatch will be executed, as the scheduledTrampolineDispatch is false, a dispatchCallback will be posted to the main thread handler;
4. Because will not call setContent, the main thread handler got no chance to execute, the suspended coroutine cannot be resumed;
5. **Once all test methods in AViewModelTest executed done, the Robolectric test environment will be reset. The main thread message queue will also be reset, the posted dispatchCallback is dropped.**
Until now, the GlobalSnapshotManager's channel is not empty, so that the global write observer will not try to resume the coroutine even state modification occurs; the AndroidUiDispatcher's scheduledTrampolineDispatch is true, so that no coroutine run in main dispatcher can be resumed.

Execute BScreenTest.kt or any other compose screen test as long as which will call setContent method
1. When calling setContent, RobolectricIdlingStrategy.runUntilIdle will be executed, Snapshot.current.hasPendingChanges() will return true since the previous state modification cannot be applied;
2. AppNotIdleException happened after 1 minute(timeout);
3. The issue happened;

Any subsequent compose screen test will get this issue.

**Comment by alwa** [comment]
I can confirm that the problems are still there in Gradle 8/AGP 8/JDK 17

**Comment by ArpitPatel009** [comment]
I'm having a similar problem. It works when we run a single file, however when we run the entire test suite, 250 tests fail. I tried adding '@org.robolectric.annotation.Config(sdk = [Build.VERSION_CODES.Q])' to one of the files, but it kept failing. 

This is happening after i update 
compose 1.0.5 -> 1.2.0
AGP 7.0.4 -> 7.4.2

Is there a workaround for this problem? We now have over 24 files that need to be fixed, and adding this to all of them appears to be a lengthy job. 

Is there any progress being made on this issue? When can we expect a solution?

I know there is enough evidence to reproduce this issue but adding logs in case there is any helpful info. FYI we also have legacy Java code and we are also using compose in our project. 



```
@RunWith(AndroidJUnit4::class)
@org.robolectric.annotation.Config(sdk = [Build.VERSION_CODES.Q])
class MainActivityLogoutTests {
    @get:Rule
    val activityScenarioRule = LazyActivityScenarioRule(createEmptyComposeRule(), MainActivity::class.java)

    @Mock
    private lateinit var serviceHelper: ServiceHelper

    @Mock
    private lateinit var chatbot: ChatbotServiceHelper

    @Mock
    private lateinit var viewModel: MainActivityViewModel

    @Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)

        ServiceHelper.setInstance(serviceHelper)
        ChatbotServiceHelper.setInstance(chatbot)

        viewModel = DaggerHelper.setViewModel(defaultAnswer = Answers.RETURNS_MOCKS) {
            on { config } doReturn Config()
        }

        activityScenarioRule.launch()
    }

    @After
    fun tearDown() {
        ServiceHelper.setInstance(null)
        ChatbotServiceHelper.setInstance(null)
        DaggerHelper.reset()
    }

    @Test
    fun logout_tellsTheServer() {
        // arrange

        // act
        activityScenarioRule.scenario.onActivity {
            it.logout()
        }

        //assert
        verify(viewModel).logout()
    }

    @Test
    fun logout_clearsCookies() {
        // arrange
        assumeTrue("Can't inject config values", BuildConfig.CHATBOT_ENABLED)

        // act
        activityScenarioRule.scenario.onActivity {
            it.logout()
        }

        // assert
        verify(chatbot).deleteCookies()
    }

    @Test
    fun logout_erasesChatbotRepo() {
        // arrange

        // act
        activityScenarioRule.scenario.onActivity {
            it.logout()
        }

        // assert
        assertThat("chatbot repo not erased",
                ChatbotRepository.realInstance, `is`(nullValue()))
    }

    @Test
    fun logout_opensLoginActivity() {
        // arrange

        // act
        activityScenarioRule.scenario.onActivity {
            it.logout()

            // assert
            val intent = shadowOf(it).nextStartedActivity
            assertThat("MainActivity didn't start an Activity",
                    intent, `is`(notNullValue()))
            assertThat("Intent is for the wrong class",
                    intent.component!!.className, `is`(LoginActivity::class.java.name))
        }
    }

    @Test
    fun logout_closesTheActivity() {
        // arrange

        // act
        activityScenarioRule.scenario.onActivity {
            it.logout()

            // assert
            assertThat("MainActivity should finish() when it opens the LoginActivity",
                    it.isFinishing, `is`(true))
        }
    }

}
````






```
Binary XML file line #48: Binary XML file line #48: Error inflating class com.google.android.material.bottomnavigation.BottomNavigationView
android.view.InflateException: Binary XML file line #48: Binary XML file line #48: Error inflating class com.google.android.material.bottomnavigation.BottomNavigationView
	Suppressed: org.robolectric.android.internal.AndroidTestEnvironment$UnExecutedRunnablesException: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(Looper.getMainLooper()).idle() call.
Caused by: android.view.InflateException: Binary XML file line #48: Error inflating class com.google.android.material.bottomnavigation.BottomNavigationView
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Unknown Source)
	at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createView(LayoutInflater.java:854)
	at android.view.LayoutInflater.createView(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createView(LayoutInflater.java:776)
	at android.view.LayoutInflater.createView(LayoutInflater.java)
	at io.github.inflationx.viewpump.internal.-ViewPumpLayoutInflater.createCustomViewInternal(-ViewPumpLayoutInflater.kt:209)
	at io.github.inflationx.viewpump.internal.-ViewPumpLayoutInflater.access$createCustomViewInternal(-ViewPumpLayoutInflater.kt:19)
	at io.github.inflationx.viewpump.internal.-ViewPumpLayoutInflater$PrivateWrapperFactory2ViewCreator.onCreateView(-ViewPumpLayoutInflater.kt:421)
	at io.github.inflationx.viewpump.internal.-FallbackViewCreationInterceptor.intercept(-FallbackViewCreationInterceptor.kt:14)
	at io.github.inflationx.viewpump.internal.-InterceptorChain.proceed(-InterceptorChain.kt:30)
	at io.github.inflationx.calligraphy3.CalligraphyInterceptor.intercept(CalligraphyInterceptor.java:18)
	at io.github.inflationx.viewpump.internal.-InterceptorChain.proceed(-InterceptorChain.kt:30)
	at io.github.inflationx.viewpump.ViewPump.inflate(ViewPump.kt:36)
	at io.github.inflationx.viewpump.internal.-ViewPumpLayoutInflater$PrivateWrapperFactory2.onCreateView(-ViewPumpLayoutInflater.kt:399)
	at android.view.LayoutInflater$FactoryMerger.$$robo$$android_view_LayoutInflater_FactoryMerger$onCreateView(LayoutInflater.java:239)
	at android.view.LayoutInflater$FactoryMerger.onCreateView(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$tryCreateView(LayoutInflater.java:1069)
	at android.view.LayoutInflater.tryCreateView(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:997)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:961)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$rInflate(LayoutInflater.java:1123)
	at android.view.LayoutInflater.rInflate(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$rInflateChildren(LayoutInflater.java:1084)
	at android.view.LayoutInflater.rInflateChildren(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:682)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at io.github.inflationx.viewpump.internal.-ViewPumpLayoutInflater.inflate(-ViewPumpLayoutInflater.kt:57)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:534)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
````

**Comment by likeprogram** [comment]
> 

One possible solution for this issue, just make sure Compose UI tests and Compose logic file tests do not use the same looper mode. Since the default looper mode is PAUSED, you can create two base classes, one for Compose UI tests which uses PAUSED looper mode, the other one for non Compose UI tests which use LEGALCY looper mode. So that the Compose UI tests and non Compose UI tests will use different Roberletric sandbox, then the issue should be fixed.

The other possible solution is use a test suite to specify the execution order of all tests, execute all Compose UI tests firstly, then all non Compose UI tests. Or execute all non Compose UI tests firstly, and then all Compose UI tests. The key point is do not mix the Compose UI test and non Compose UI test.

Hope these solutions can solve your problem.

**Comment by ArpitPatel009** [comment]
Thank you for quick prompt. Is there any snippet I can take a look as I have no idea what is Looper and how it works lol 😄

**Comment by likeprogram** [comment]
> Thank you for quick prompt. Is there any snippet I can take a look as I have no idea what is Looper and how it works lol 😄

You can use Roberletric Config annotation to specify which looper mode you want to use, use a different looper mode can avoid Roberletric sandbox reusing between Compose UI test and non Compose UI test.

**Comment by ArpitPatel009** [comment]
I tried adding `@LooperMode(LooperMode.Mode.LEGACY)` to above test file and It is complaining about viewModel that Dagger generate on `onCreate`. 

It might not be related to this as we have depreciated code in our MainActivity. 

```
viewModel = new ViewModelProvider(this, viewModelFactory).get(MainActivityViewModel.class);

        CustomerDetailsViewModel customerDetailsViewModel =
                new ViewModelProvider(this, viewModelFactory).get(CustomerDetailsViewModel.class);
        customerDetailsViewModel.reload();   
```


```
Cannot invoke "ca.manulife.MobileBanking.viewmodels.LiveData.CustomerDetailsViewModel.reload()" because "customerDetailsViewModel" is null
java.lang.NullPointerException: Cannot invoke "ca.manulife.MobileBanking.viewmodels.LiveData.CustomerDetailsViewModel.reload()" because "customerDetailsViewModel" is null
	at ca.manulife.MobileBanking.MainActivity.setUpViewModels(MainActivity.java:419)
	at ca.manulife.MobileBanking.MainActivity.onCreate(MainActivity.java:202)
	at android.app.Activity.performCreate(Activity.java:8051)
	at android.app.Activity.performCreate(Activity.java:8031)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1329)
	at org.robolectric.android.internal.RoboMonitoringInstrumentation.callActivityOnCreate(RoboMonitoringInstrumentation.java:284)
	at org.robolectric.android.controller.ActivityController.lambda$create$0(ActivityController.java:113)
	at org.robolectric.shadows.ShadowLegacyLooper.runPaused(ShadowLegacyLooper.java:334)
	at org.robolectric.android.controller.ActivityController.create(ActivityController.java:113)
	at org.robolectric.android.controller.ActivityController.create(ActivityController.java:118)
```

**Comment by ChrisAJS** [comment]
Building off of @likeprogram 's deep dive, I managed to get what would have been a failing test, to pass by executing the pending dispatch callback:

```kotlin
    private fun environmentIsSane() {
        val clazz =
            javaClass.classLoader.loadClass("androidx.compose.ui.platform.AndroidUiDispatcher")
        val combinedContextClass =
            javaClass.classLoader.loadClass("kotlin.coroutines.CombinedContext")
        val companionClazz = clazz.getDeclaredField("Companion").get(clazz)
        val combinedContext = companionClazz.javaClass.getDeclaredMethod("getMain")
            .invoke(companionClazz) as CoroutineContext

        val androidUiDispatcher = clazz.cast(
            combinedContextClass.getDeclaredField("element").apply { isAccessible = true }
                .get(combinedContext)
        )

        var scheduledFrameDispatch =
            clazz.getDeclaredField("scheduledFrameDispatch").apply { isAccessible = true }
                .getBoolean(androidUiDispatcher)
        var scheduledTrampolineDispatch =
            clazz.getDeclaredField("scheduledTrampolineDispatch").apply { isAccessible = true }
                .getBoolean(androidUiDispatcher)

        val dispatchCallback = clazz.getDeclaredField("dispatchCallback").apply { isAccessible = true }.get(androidUiDispatcher) as Runnable

        if (scheduledFrameDispatch || scheduledTrampolineDispatch) {
            dispatchCallback.run()
            scheduledFrameDispatch =
                clazz.getDeclaredField("scheduledFrameDispatch").apply { isAccessible = true }
                    .getBoolean(androidUiDispatcher)
            scheduledTrampolineDispatch =
                clazz.getDeclaredField("scheduledTrampolineDispatch").apply { isAccessible = true }
                    .getBoolean(androidUiDispatcher)
        }

        assertFalse(scheduledFrameDispatch)
        assertFalse(scheduledTrampolineDispatch)
    }
```

Unfortunately, I can't find a nice place to place this in due to the nature of the sandbox availability between test methods. I attempted to place it in a subclass of the `RobolectricTestRunner`, but none of the places I attempted to place it would allow me to get around the `Looper` class not being mocked.

I think I'm going to create an extension on the `ComposeContentTestRule` class to provide a `safeSetContent` method instead.

My understanding of Robolectric and how it actually accomplishes the android framework class replacement at test runtime is likely letting me down here.

**Comment by vitusortner** [comment]
We're running into the same problem and interestingly, the errors only appear in one module, which runs both Compose UI tests and Android view model tests – nothing unordinary. Additionally, the tests only fail when executing all tests of the package. Executing a previously failing test class doesn't fail when only the test class is run. The error message is attached below.

We're able to work around the issue by adding `androidx.loader.content` to the instrumented packages.

```kotlin
@Config(instrumentedPackages = ["androidx.loader.content"])
```

Robolectric: 4.9.2 (same issue with 4.10.3)
Compose BOM: 2023.05.01 (same issue with 2023.06.01)
Compose compiler: 1.4.7

```
Compose did not get idle after 4259511 attempts in 60 SECONDS. Please check your measure/layout lambdas, they may be causing an infinite composition loop. Or set Espresso's master idling policy if you require a longer timeout. The following Idle Conditions failed .
androidx.test.espresso.AppNotIdleException: Compose did not get idle after 4259511 attempts in 60 SECONDS. Please check your measure/layout lambdas, they may be causing an infinite composition loop. Or set Espresso's master idling policy if you require a longer timeout. The following Idle Conditions failed .
	at androidx.test.espresso.AppNotIdleException.create(AppNotIdleException.java:11)
	at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:55)
	at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy$runUntilIdle$1.invoke(RobolectricIdlingStrategy.android.kt:48)
	at androidx.compose.ui.test.junit4.AndroidSynchronization_androidKt.runOnUiThread(AndroidSynchronization.android.kt:33)
	at androidx.compose.ui.test.junit4.RobolectricIdlingStrategy.runUntilIdle(RobolectricIdlingStrategy.android.kt:48)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.waitForIdle(ComposeUiTest.android.kt:344)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.access$waitForIdle(ComposeUiTest.android.kt:234)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$AndroidTestOwner.getRoots(ComposeUiTest.android.kt:561)
	at androidx.compose.ui.test.TestContext.getAllSemanticsNodes$ui_test_release(TestOwner.kt:94)
	at androidx.compose.ui.test.SemanticsNodeInteraction.fetchSemanticsNodes$ui_test_release(SemanticsNodeInteraction.kt:62)
	at androidx.compose.ui.test.SemanticsNodeInteraction.fetchOneOrDie(SemanticsNodeInteraction.kt:128)
	at androidx.compose.ui.test.SemanticsNodeInteraction.assertExists(SemanticsNodeInteraction.kt:120)
	at androidx.compose.ui.test.SemanticsNodeInteraction.assertExists$default(SemanticsNodeInteraction.kt:119)
	at com.soundcloud.android.wear.playhistory.ui.PlayHistoryScreenKtTest$should show success state$1.invoke(PlayHistoryScreenKtTest.kt:50)
	at com.soundcloud.android.wear.playhistory.ui.PlayHistoryScreenKtTest$should show success state$1.invoke(PlayHistoryScreenKtTest.kt:47)
	at com.soundcloud.android.wear.ComposeTestKt.runComposeTest(ComposeTest.kt:20)
	at com.soundcloud.android.wear.playhistory.ui.PlayHistoryScreenKtTest.should show success state(PlayHistoryScreenKtTest.kt:47)
	at java.base@11.0.19/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.19/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.19/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.19/java.lang.reflect.Method.invoke(Method.java:566)
	at app//org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at app//org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at app//org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at app//org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at app//org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at androidx.compose.ui.test.junit4.AndroidComposeTestRule$apply$1$evaluate$1.invoke(AndroidComposeTestRule.android.kt:272)
	at androidx.compose.ui.test.junit4.AndroidComposeTestRule$apply$1$evaluate$1.invoke(AndroidComposeTestRule.android.kt:271)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$AndroidComposeUiTestImpl.withDisposableContent(ComposeUiTest.android.kt:512)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1$1$1$1$1$1.invoke(ComposeUiTest.android.kt:330)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.withTextInputService(ComposeUiTest.android.kt:396)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.access$withTextInputService(ComposeUiTest.android.kt:234)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1$1$1$1$1.invoke(ComposeUiTest.android.kt:329)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.withComposeIdlingResource(ComposeUiTest.android.kt:383)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.access$withComposeIdlingResource(ComposeUiTest.android.kt:234)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1$1$1$1.invoke(ComposeUiTest.android.kt:328)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.withWindowRecomposer(ComposeUiTest.android.kt:357)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.access$withWindowRecomposer(ComposeUiTest.android.kt:234)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1$1$1.invoke(ComposeUiTest.android.kt:327)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.withTestCoroutines(ComposeUiTest.android.kt:370)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.access$withTestCoroutines(ComposeUiTest.android.kt:234)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1$1.invoke(ComposeUiTest.android.kt:326)
	at androidx.compose.ui.test.junit4.IdlingStrategy.withStrategy(IdlingStrategy.android.kt:52)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1$1.invoke(ComposeUiTest.android.kt:325)
	at androidx.compose.ui.test.junit4.IdlingResourceRegistry.withRegistry(IdlingResourceRegistry.jvm.kt:157)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment$runTest$1.invoke(ComposeUiTest.android.kt:324)
	at androidx.compose.ui.test.junit4.ComposeRootRegistry.withRegistry(ComposeRootRegistry.android.kt:146)
	at androidx.compose.ui.test.AndroidComposeUiTestEnvironment.runTest(ComposeUiTest.android.kt:323)
	at androidx.compose.ui.test.junit4.AndroidComposeTestRule$apply$1.evaluate(AndroidComposeTestRule.android.kt:271)
	at app//org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at app//org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:580)
	at app//org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:287)
	at app//org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:99)
```

**Comment by dbreese** [comment]
This issue went away for me by changing `LaunchedEffect(Unit) {...}` to `LaunchedEffect(true) {...}`.

Of course, this may have other side effects.

**Comment by dennismozart1994** [comment]
@ChrisAJS  would you be able to provide a snippet on how that extension would look like? That'd be pretty helpful for everyone having issues with compose like this annoying exception.

**Comment by ChrisAJS** [comment]
> @ChrisAJS would you be able to provide a snippet on how that extension would look like? That'd be pretty helpful for everyone having issues with compose like this annoying exception.

Apologies for the late reply.

The idea with the `setSafeContent {}` extension is to replace your use of `composeTestRule.setContent {}`.

This `setSafeContent {}` call would purge the schedulers so that the app will appear idle to the testing frameworks.

**Comment by johngray1965** [comment]
What fixed the problem for me was creating my own Junit Rule:
`class ResetChoreographerRule : TestRule {
    override fun apply(
        base: Statement?,
        description: Description?
    ): Statement =
        object : Statement() {
            override fun evaluate() {
                try {
                    val uiDispatcher: AndroidUiDispatcher = ReflectionHelpers.getField(AndroidUiDispatcher.Main, "element")
                    ReflectionHelpers.setField(uiDispatcher, "choreographer", Choreographer.getInstance())
                } catch (e: Exception) {
                    Timber.e(e, "Failed to reset Choreographer")
                }
                try {
                    base?.evaluate()
                } finally {
                    // Reset the Choreographer after the test
                }
            }
        }
}
`
And making sure its before the compose rule.  I was seeing tests fail setting up the compose rule.

**Comment by brettchabot** [comment]
I believe this issue is a duplicate of #9043 . Please give the next release of Robolectric a try

**Comment by frett** [comment]
I'm still getting `androidx.test.espresso.AppNotIdleException` when I upgrade robolectric to `4.14-beta-1` and remove the workaround mentioned in https://github.com/robolectric/robolectric/issues/7055#issuecomment-1551119229

As I was working through where best to apply the workaround in my own project I found that I only needed to run it after my [Circuit](https://github.com/slackhq/circuit) presenter tests. Internally the Circuit presenter tests use [molecule](https://github.com/cashapp/molecule) which has it's own copy of the `AndroidUiDispatcher`. But I only have to apply the workaround to the `AndroidUiDispatcher` that comes with Compose UI, so I'm not sure if the molecule copy matters or is just a red herring.

**Comment by fraune** [comment]
I think I am also having this issue. I get the same "Compose did not get idle after ..." message.

Luckily, [`@Config(sdk = [Build.VERSION_CODES.Q])`](https://github.com/robolectric/robolectric/issues/7055#issuecomment-1077577227) still seems to be an effective workaround.

Our app is transitioning to Compose, so we set the Compose view content from a fragment, as suggested [in the Android docs](https://developer.android.com/develop/ui/compose/migrate/interoperability-apis/compose-in-views). These types of views seem to be the ones that I most frequently get this error on. Not when run individually, but most often as part of the full UI test suite, or a larger/mixed test class.
```xml
<androidx.compose.ui.platform.ComposeView
    android:id="@+id/help_button"
    ... />
```
```kotlin
override fun onCreateView(...) {
    requireBinding().helpButton.setContent {
        HelpButton()
    }
}
```
